### PR TITLE
Support delete from in clickhouse connector

### PR DESCRIPTION
**Pull request by sahoss**
## Description

1. Bumps clickhouse and altinity images used in tests. Also update the recommended version in docs.
2. Supoprt delete in clickhouse connector by removing overridden delete() method. Update associated tests.

## Additional context and related issues
https://github.com/trinodb/trino/issues/18067 

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
(x) Release notes are required, with the following suggested text:

```markdown
# Section
* Add support for delete in clickhouse connector. ({issue}`18067`)
```

### CODE REVIEW

**Comment by ebyhr** [comment]
What's the difference from #18550?

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java [hunk 1]
```diff
@@ -386,7 +386,7 @@ public void testDouble()
         SqlDataTypeTest.create()
                 .addRoundTrip("double", "3.1415926835", DOUBLE, "DOUBLE '3.1415926835'")
                 .addRoundTrip("double", "1.79769E308", DOUBLE, "DOUBLE '1.79769E308'")
-                .addRoundTrip("double", "2.225E-307", DOUBLE, "DOUBLE '2.225E-307'")
+                .addRoundTrip("double", "2.2250738585072014E-308", DOUBLE, "DOUBLE '2.2250738585072014E-308'")
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java:389]: Why did you change this value?

Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java:389]: there were too many merge conflicts/that branch was stale. adding the same changes into this PR

**Comment by sahoss** [comment]
> What's the difference from #18550?


there were too many merge conflicts/that branch was stale. adding the same changes into this PR

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestAltinityConnectorSmokeTest.java [hunk 2]
```diff
@@ -34,13 +32,4 @@ protected QueryRunner createQueryRunner()
                 ImmutableMap.of("clickhouse.map-string-as-varchar", "true"), // To handle string types in TPCH tables as varchar instead of varbinary
                 REQUIRED_TPCH_TABLES);
     }
-
-    @Test
-    @Override
-    public void testRenameSchema()
```
Comment (sahoss) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestAltinityConnectorSmokeTest.java:40]: rename is supported in latest version (23.3)

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java [hunk 3]
```diff
@@ -53,4 +55,46 @@ public void testShowCreateTable()
...
+    {
+        // Override because default storage engine doesn't support delete
+        String tableDefinition = "(" +
+                " regionkey BIGINT NOT NULL," +
+                " name VARCHAR(25) NOT NULL," +
+                " comment VARCHAR(115) NOT NULL" +
+                "    ) WITH (engine='MergeTree',  order_by=ARRAY['regionkey'])";
+
+        try (TestTable table = new TestTable(getQueryRunner()::execute, "test_delete_all_data", tableDefinition)) {
+            assertQuerySucceeds(format("insert into %s select * from region", table.getName()));
```
Comment (sahoss) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java:71]: not possible create table with "as select * from region".  default  clickhouse engine in test(log) does not support delete

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java (hunk 1 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java:389]: somethings changed in clickhouse 22.8 -> 22.9. i am unable to point at what. pasting the 22.8 and 22.9 results for the same 3 queries

```sql
 create table myTable( col1 Float32, col2 Float64) ENGINE=MergeTree() ORDER BY col1
insert into myTable VALUES (2.225E-307, 2.225E-307);
select * from myTable
```

**22.9**
image



**22.8**
image

**Comment by sahoss** [comment]
@ebyhr this PR can now be reviewed. thank you.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java (hunk 1 above)
Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java:389]: Please file an issue to ClickHouse and leave the link here. 

Also, please add a test verifying the correctness instead of removing the test so that we can catch the change of ClickHouse.

Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseTypeMapping.java:389]: https://github.com/ClickHouse/ClickHouse/issues/56381.

@ebyhr i have added the testcases that will catch a fix(once made my clickhouse)

**Comment by sahoss** [comment]
@ebyhr checking to see if you can review this pr. thanks.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java [hunk 4]
```diff
@@ -552,8 +552,8 @@ public boolean isLimitGuaranteed(ConnectorSession session)
     @Override
     public OptionalLong delete(ConnectorSession session, JdbcTableHandle handle)
     {
-        // ClickHouse does not support DELETE syntax, but is using custom: ALTER TABLE [db.]table [ON CLUSTER cluster] DELETE WHERE filter_expr
-        throw new TrinoException(NOT_SUPPORTED, MODIFYING_ROWS_MESSAGE);
+        checkArgument(!handle.getConstraint().isAll(), "Deleting without constraint is not supported: %s", handle);
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java:555]: We don't need to disallow this condition because ClickHouse supports `WHERE true` condition.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java [hunk 5]
```diff
@@ -53,4 +55,46 @@ public void testShowCreateTable()
                         "   engine = 'LOG'\n" +
                         ")");
     }
+
+    @Override
+    @Test
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java:60]: nit: Replace the order for consistency with other places.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java [hunk 6]
```diff
@@ -53,4 +55,46 @@ public void testShowCreateTable()
...
+    public void testDeleteAllDataFromTable()
+    {
+        // Override because default storage engine doesn't support delete
+        String tableDefinition = "(" +
+                " regionkey BIGINT NOT NULL," +
+                " name VARCHAR(25) NOT NULL," +
+                " comment VARCHAR(115) NOT NULL" +
+                "    ) WITH (engine='MergeTree',  order_by=ARRAY['regionkey'])";
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java:68]: Is it possible to use CTAS statement instead?

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java (hunk 3 above)
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java:71]: Uppercase `insert into ... select from`.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java [hunk 7]
```diff
@@ -29,14 +29,14 @@
...
+    private static final DockerImageName CLICKHOUSE_IMAGE = DockerImageName.parse("clickhouse/clickhouse-server");
+    public static final DockerImageName CLICKHOUSE_LATEST_IMAGE = CLICKHOUSE_IMAGE.withTag("23.8.4.69");
+    public static final DockerImageName CLICKHOUSE_DEFAULT_IMAGE = CLICKHOUSE_IMAGE.withTag("23.3.8.21"); // EOL is Mar 2024
 
     // Altinity Stable Builds Life-Cycle Table https://docs.altinity.com/altinitystablebuilds/#altinity-stable-builds-life-cycle-table
     private static final DockerImageName ALTINITY_IMAGE = DockerImageName.parse("altinity/clickhouse-server").asCompatibleSubstituteFor("yandex/clickhouse-server");
-    public static final DockerImageName ALTINITY_LATEST_IMAGE = ALTINITY_IMAGE.withTag("21.8.13.1.altinitystable");
-    public static final DockerImageName ALTINITY_DEFAULT_IMAGE = ALTINITY_IMAGE.withTag("20.8.4.11_aes"); // EOL is 02 December 2022
+    public static final DockerImageName ALTINITY_LATEST_IMAGE = ALTINITY_IMAGE.withTag("23.3.13.7.altinitystable");
+    public static final DockerImageName ALTINITY_DEFAULT_IMAGE = ALTINITY_IMAGE.withTag("23.3.13.7.altinitystable"); // EOL is 14 July 2026
```
Comment (ebyhr) [inline plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java:39]: Altinity 21.3 is still supported https://docs.altinity.com/altinitystablebuilds/#altinity-stable-builds-life-cycle-table

**Review by ebyhr — CHANGES_REQUESTED** [review]

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java (hunk 4 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: removed override

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java (hunk 5 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java]: changed

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java (hunk 3 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java]: changed everywhere

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java (hunk 7 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java]: updated to use 21.3

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java (hunk 6 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/BaseClickHouseConnectorSmokeTest.java:68]: can you please give an example how @ebyhr ? (while being able to specify the table properties)

the issue i am running into is like this
1. the tables can be created but the column will all nullable. 
2. clickhouse requires atleast one non null column for the order_by

PS: 2 can be overcome by a clickhouse property that allows nullable order_by keys, however its not recognized in trino yet https://clickhouse.com/docs/en/operations/settings/settings#allow-nullable-key

**Comment by ebyhr** [comment]
Please confirm CI failure.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java (hunk 7 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java]: @ebyhr 21.3 does not support delete. this was causing some of the tests to fail. 23.3 is the only long term supported version that also support delete.

delete was introduced in [[22.9] ](https://clickhouse.com/docs/en/whats-new/changelog/2022#-clickhouse-release-229-2022-09-22)

https://docs.altinity.com/altinitystablebuilds/#altinity-stable-builds-life-cycle-table

Comment (ebyhr) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java]: I know. Please verify the failure in Altinity test class.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java (hunk 4 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: @ebyhr jfyi: without this check in trino, the error returned to the user is not user friendly. (the error from clickhouse is propagated back to the trino user). it makes sense in the clickhouse context, but not as per the format of other trino errors.

Comment (ebyhr) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: You need to modify DELETE query to include `WEHRE true` instead.

File: plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java (hunk 7 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/test/java/io/trino/plugin/clickhouse/TestingClickHouseServer.java:39]: @ebyhr : its not just about the test cases. some of the behavior changes too between 21.3 and 23.3 (eg: IPv4/ipv6 mapping logic). also the tests themselves will end up split  - because 21.3 does not support delete whereas 23.3 does. so we will end up having two sets of tests+code. before i make this change, just confirming this was what you meant.

File: plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java (hunk 4 above)
Comment (sahoss) [reply plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java]: @ebyhr - sorry i did not follow this. are you suggesting to add the   where true clause in the Clickhouse client if no condition is mentioned in the trino query?

**Comment by mosabua** [comment]
:wave: @sahoss - this PR has become inactive. We hope you are still interested in working on it. Please let us know, and we can try to get reviewers to help with that.

We're working on closing out old and inactive PRs, so if you're too busy or this has too many merge conflicts to be worth picking back up, we'll be making another pass to close it out in a few weeks.
